### DISCUSSION TITLE
Continuous Deployment pipeline using Argo

### DISCUSSION THREAD

**Discussion by NarsGNA** [juspay/hyperswitch#4108] [General]
# Context
As an organization committed to automating application deployment and lifecycle management, the primary goal is to implement a Continuous Deployment (CD) pipeline using Argo. This solution aims to provide automation, auditability, and ease of understanding throughout the application's lifecycle.

# Expected outcomes
## Argo CD Integration

- Implement Argo CD as the primary tool for continuous deployment, ensuring that it is seamlessly integrated into the existing development workflow.

- Utilize Argo CD to manage the application deployment lifecycle, including syncing with the desired state defined in Git repositories.

## Automation

- Automate the deployment process using Argo CD to eliminate manual intervention and reduce the risk of human errors.

- Integrate automated testing, such as unit tests and integration tests, into the CD pipeline to validate the correctness of each deployment.

## Infrastructure as Code (IaC)

- Leverage Infrastructure as Code principles to define and manage the infrastructure configuration needed for the application deployment.

- Store infrastructure configuration in version-controlled repositories alongside application code to maintain consistency and traceability.

## Auditability

- Implement logging and auditing mechanisms within Argo CD to capture detailed information about each deployment, including changes made, success/failure status, and who initiated the deployment.

- Ensure that audit logs are securely stored and easily accessible for compliance and troubleshooting purposes.

## Rollout and Rollback Strategies

- Define and implement rollout strategies within Argo CD to control the pace and impact of application updates, allowing for gradual deployments to specific environments.

- Establish a reliable rollback mechanism in Argo CD to quickly revert to the previous state in case of deployment issues.

## Environment Configuration

- Utilize Argo CD to manage environment-specific configurations, ensuring consistency across development, staging, and production environments.

- Allow for dynamic configuration adjustments through Argo CD to accommodate variations in deployment environments.

## Notification and Reporting

- Implement notification mechanisms within Argo CD to alert relevant stakeholders about the status of deployments.
Generate deployment reports that summarize key information, including success/failure rates, deployment durations and any post-deployment actions taken.

## Ease of Understanding

- Provide clear and concise documentation for the Argo CD setup, configuration, and usage.

- Conduct training sessions to ensure that team members have a solid understanding of the CD pipeline and can effectively use Argo CD for deployment and lifecycle management.

# How to get started
- Start your local kubernetes(minikube)
- Use our [helm chart](https://github.com/juspay/hyperswitch-helm) to deploy the stack in your local kubernetes (checkout ReadME for more details)
- Start with the idea for the solution (Attach it with the PR)
- Raise PR in [hyperswitch Helm chart](https://github.com/juspay/hyperswitch-helm) repo

**Comment by varshith257** [comment]
@NarsGNA I am also quite interested in working on it. I have experience with Argo. Can clarify what the hiring hiring-challenge label is? What could be my next steps?

**Reply by sanjaymenaria03** [reply]
they are hiring on the basis of open source contribution, you can checkout their career page for more info.

**Reply by neerajkumar-n** [reply]
Yup! You can check out our Career page in LinkedIn[https://www.linkedin.com/company/hyperswitch/jobs] or you can check our Wiki [https://github.com/juspay/hyperswitch/wiki/Hyperswitch-Hiring-Challenge] to know more

**Comment by kitswas** [comment]
@NarsGNA I am interested in working on this problem. How should I get started?

**Reply by neerajkumar-n** [reply]
@kitswas - we have updated the discussion to give more info regarding this. Do take a look at this.

**Comment by arpit-pi** [comment]
@NarsGNA Interested, would like to work on this. How should i get started on this?

**Reply by inventvenkat** [reply]
We have edited the description about `how to get started` please check it out.

**Comment by lucifer1708** [comment]
@NarsGNA 
I am interested.
Do i have to deploy it on my own server using my  own fork?

**Reply by inventvenkat** [reply]
Yes, you have to fork and work. Once the work is done, Raise a PR. 
Also please check the `How to get started` in the description, Which is been added recently.

**Comment by aialok** [comment]
I would love to work on this issue. I have started working on this. 
Thank u : )

**Reply by neerajkumar-n** [reply]
@aialok - sure.

**Comment by abhinavtyagiO** [comment]
Hi! I am interested in contributing towards this. Can anyone tell if there's a deadline for this?

**Reply by neerajkumar-n** [reply]
@abhinavtyagiO  - I wouldn't say that there is a deadline, but would recommend you to contribute at the earliest so we can evaluate and reach out to the right candidates.

**Comment by RohitRathore1** [comment]
@NarsGNA & @c10n3 I want to know that we do not have any running k8s cluster/defined services & deployment already in this directory?

**Reply by neerajkumar-n** [reply]
@RohitRathore1 - if you refer to "How to get started section". It should answer your question.

We did make some updates on the problem statements to give more clarity around this.

**Reply by RohitRathore1** [reply]
@neerajkumar-n Thanks! Just for the clarity, now while configuring and creating this app in ArgoCD [this path](https://github.com/juspay/hyperswitch-helm/tree/main/charts/incubator/hyperswitch-app) we will provide to the `--path` placeholder?

**Reply by RohitRathore1** [reply]
@NarsGNA @neerajkumar-n 

> Implement logging and auditing mechanisms within Argo CD to capture detailed information about each deployment, including changes made, success/failure status, and who initiated the deployment.

Can you point out any supporting material around it? As per my research I did not find that Argo CD support something like this.

**Reply by jagan-jaya** [reply]
@RohitRathore1 Please use [hyperswitch-stack](https://github.com/juspay/hyperswitch-helm/tree/main/charts/incubator/hyperswitch-stack) as it has all the underlying hyperswitch services

**Comment by Parvej2001** [comment]
Hello hiring team , Thank you for the challenge. I am interested and I have started working on it.

**Reply by neerajkumar-n** [reply]
Great, we have updated on how to get started. Do take a look

**Comment by THIRUMALAIKUMAR009** [comment]
https://github.com/THIRUMALAIKUMAR009/https-github.com-juspay-hyperswitch-helm.git

**Comment by Shubh-k04** [comment]
@NarsGNA @neerajkumar-n I've submitted my proposal for this task!

**Comment by hold7door** [comment]
Hi team,
Kind-of learning ArgoCD and Kubernetes along the way with this.

Basically my approach would be as follows-

1. Setup ArgoCD to monitor the hyperswitch-helm repo on the necessary path.
2. Whenever there is commit pushed to the 'hyperswitch' repo ,github action is triggered which updates the docker image version in hyperswitch-helm repo. ArgoCD will detect this change and trigger deployment.

**Comment by Ovank** [comment]
@NarsGNA , Is this challenge still up from Hiring challenge. I could see two PR for this implementation.

**Comment by nike541** [comment]
@NarsGNA , Hello hiring team , Thank you for the challenge. I am interested and I have started working on it.

**Reply by SAGARGAUD01** [reply]
Hello Hiring Team ,Thank you for the opportunity I am going ahead with this task.
